README.md
# data-center

## usage
本项目采用flask，主要用到了os，re，pandas，requests，typing等库。
python版本建议3.6及以上，直接运行下面的命令即可执行后端
- pip install -r requirements.txt
- python main.py
注：
1. requirements.txt中的包可能不全，我因为没用venv，freeze出来手动筛了一遍，有什么缺失的库用pip install安装即可。
2. 文件路径只有部分使用了os.path.join。

## 设计思路
### 文件（夹）介绍
接下来我对几个重要文件及文件夹进行简单介绍。
- oerv_obsdata 文件夹克隆自[OBS构建数据仓库](https://gitee.com/phoebe-xi/oerv_obsdata)，本项目目前几乎所有数据都是取自此处。
- oerv_script 内主要要两个文件夹，分别是oerv_obsdata和script文件夹，其中oerv_obsdata文件夹同上，但是用途有所区别。因为本项目有一个一键更新数据仓库的功能，所以在将该oerv_obsdata克隆到本地后，执行数据更新代码，再将其推送到gitee，总而言之，其是作为一个中间文件夹存在的；script克隆自[OBS数据抓取仓库](https://gitee.com/phoebe-xi/oerv_script/tree/master)，修改了部分代码，详情可见oerv_script的README。
- changelog.md 文件主要记录版本变动，包括一些需求和代码设计的描述。
- main.py是flask的入口文件，在执行后端的时候执行“python main.py”。
- ObservatoryPackages.py文件存放网页端Package Observatory分区相关的api。
- ObservatoryProjects.py文件存放网页端Project Observatory分区相关的api。
- Update.py文件主要用于更新数据，包括1. 只从远程拉取文件到oerv_obsdata；2. 更新远程的oerv_obsdata仓库并拉取数据到本地。其中第二项功能可能需要几分钟时间。
- Utils.py中包含提取文件差异的代码。可以通过commit id, path等参数对所需要比较的文件进行配置。

### 如何新增一个页面对应的api
参考ObservatoryPackages.py等文件创建一个新文件，并在main.py中注册蓝图即可。


changelog.md
# v0.3.0
- 工程比较，典型应用场景：
1. 修包：修复在 2203中构建成功，但是2209工程中构建失败的包——》输出哪些包2203成功，但是2209失败
2. 对比：同一个工程，不同的仓库下的构建差异。Java工程 2203仓库和2209仓库没有构建成功的包对比
3. 进展观察：同一个工程，同一个仓库，上个月和这个月的差异，哪些包原来失败，现在成功


- 新增ObservatoryPackages.py文件，将原本的misc.py文件中的包搜索功能代码移入该文件，添加相似工程差异比较方面代码。

  - 新增两个页面的功能分别从时间（Temporal）和空间（Spatial）两个维度对相似工程进行比较。
  - 时间维度，比较单个仓库在不同时间的差异。

    - 为了实现这个功能，我将之前比较项目的统计数据差异的代码提取成匹配函数match_ordered、match_unordered和一个类Differ，存放于Utils.py文件中。
    - 通过改变Differ的成员变量，例如commit id和path，即可实现不同数据格式、不同路径下文件的差异比较，使得Differ类的差异比较功能具有较高泛用性。
    - match_ordered函数对有序数据进行差异比较，使用的方法是循环遍历。
    - match_unordered函数对无序数据进行差异比较，使用hash和集合(交集，差集)的方法切割数据。
  
  - 空间维度，比较最新数据下两个不同仓库的差异。差异比较方法用的是match_unordered

# v0.2.0
- 后端从Django向Flask迁移
  
  考虑Django文件组织较为复杂，而且有大多数功能是用不上的，因此向更为轻便的Flask迁移，以期提高开发效率。
- flask app模块化

  原先的代码将所有的接口都放在一个文件中，不利于后续维护，因此通过flask的Blueprint重新组织文件，将网页上不同的Section所对应的接口划分到不同的app中。

- 优化数据更新逻辑，便于部署

  原先数据的更新只有中间文件更新，而且git pull操作散落在各个接口中，对于部署显然是不合理的。因此本次将所有和数据更新有关的git操作集中到一个函数Update.py文件的updateLocal中，再额外添加一个用于更新oerv_obsdata仓库的函数updateRepository，从而将数据更新操作集中的一个文件中进行管理。

- 更新oerv_obsdata

  Update.py中的updateRepository函数可以实现oerv_obsdata仓库的更新，实现步骤如下：
  - 将[爬取数据的仓库](https://gitee.com/phoebe-xi/oerv_script/tree/master/obs)克隆到本项目中
  - 克隆oerv_obsdata仓库到该文件夹下，并[用ssh关联到gitee](https://todebug.com/tips/)
  - 修改constant.py文件
  - 修改main.py文件dataPath(相对路径从'../..'修改成'../')
  - 在main.py文件中用os.system函数进行数据的上传。
  
  该函数在执行时首先会pull拉取数据，然后再执行抓取数据的脚本。

- Package搜索功能

  主要通过OBS API实现，调用search API，获得返回的网页，用正则表达式过滤出包和路径即可。

- 可视化图表的数据处理

  本次添加的两个图表的处理都相对枯燥繁琐，并无特殊之处。

Misc.py
from flask import Blueprint, request, jsonify

import re 

import requests 

appMisc = Blueprint('appMisc', __name__)

@appMisc.route('/search-package', methods=['POST'])
def searchPackage():
    '''
    :return [(projectName, packageName), ...]
    '''
    def constructURL(target):
        base_url = 'https://build.tarsier-infra.com/search?'
        args = 'utf8=✓&search_text={}&search_for=0&name=1&attrib_type_id='.format(target)
        return base_url + args

    # get html text
    target = request.values.get('packageName', 'gcc')
    account = {'username': 'cyanic', 'password': '123456'}
    response = requests.get(constructURL(target), auth=requests.auth.HTTPBasicAuth(**account))
    text = response.text 

    # filter out all related path in form of [(projectName, packageName), ...]
    raw = re.findall(r'package .*?title="(.*?)".*?title="(.*?)"', text)
    with open('./oerv_obsdata/projectsNames.txt') as f:
        projectsNames = f.read().split('\n')
    res = list(filter(lambda name: name[0] in projectsNames, raw))
    res = [{'projectName': item[0], 'packageName': item[1]} for item in res]
    return jsonify({'data': res})

Utils.py
from typing import Callable, List, Tuple, Dict
import os


def match_ordered(seq1: List, seq2: List, compare: Callable) -> Dict[str, List]:
    '''
    Given two ordered sequence, find the matching items and remaining elements based on compare function
    Just using two loop.
    '''
    matched1, matched2 = [], []
    mark = len(seq2) - 1
    for i in range(len(seq1) - 1, -1, -1):
        for j in range(mark, -1, -1):
            if compare(seq1[i], seq2[j]):
                matched1.append(seq1.pop(i))
                matched2.append(seq2.pop(j))
                mark = j - 1
                break
    return {'unique1': seq1, 'unique2': seq2, 'matched1': matched1, 'matched2': matched2}


def match_unordered(seq1: List, seq2: List, compareKeys: str) -> Dict[str, List]:
    '''
    Given two unordered sequences, segment them into unique and matching segmentation based on compareKeys
    Here I combine the values which correspond to the compareKeys and hash them.
    Based on hash value, using set methods to fastly segment data.
    '''
    # construct hash map
    hash1 = {hash(''.join([str(item[compareKey])
                  for compareKey in compareKeys])): item for item in seq1}
    hash2 = {hash(''.join([str(item[compareKey])
                  for compareKey in compareKeys])): item for item in seq2}

    # get hash value intersection
    hashValues1 = set(hash1.keys())
    hashValues2 = set(hash2.keys())
    intersection = hashValues1.intersection(hashValues2)

    # segment data based on intersection and difference of hash value
    unique1 = [hash1[hashValue]
               for hashValue in hashValues1.difference(hashValues2)]
    unique2 = [hash2[hashValue]
               for hashValue in hashValues2.difference(hashValues1)]
    matched1, matched2 = [], []
    for hashValue in intersection:
        matched1.append(hash1[hashValue])
        matched2.append(hash2[hashValue])
    return {'unique1': unique1, 'unique2': unique2, 'matched1': matched1, 'matched2': matched2}


class Differ:
    def __init__(self, id1: str, id2: str, sortKey: Callable, formatter: Callable, ordered: bool, comparer, filePath: str, repositoryPath: str, basePath: str):
        '''
        When ordered is True, comparer is Callable type
        When ordered is False, comparer is List[str] type 
        '''
        self.id1 = id1
        self.id2 = id2
        self.sortKey = sortKey
        self.formatter = formatter
        self.comparer = comparer
        self.ordered = ordered
        self.filePath = filePath
        self.repositoryPath = repositoryPath
        self.basePath = basePath

    def gitDiff(self):
        os.chdir(self.repositoryPath)
        os.system(
            'git diff {} {} {} > {}'.format(self.id1, self.id2, self.filePath, './diff.txt'))
        # read data to variables deleted and added
        deleted: List = []
        added: List = []
        with open('./diff.txt') as f:
            lines = f.readlines()
            # start from 6th line
            for i in range(6, len(lines)):
                if lines[i][0] == '-':
                    # strip off + / - and new line character
                    deleted.append(lines[i][1:-1].split('  '))
                elif lines[i][0] == '+':
                    added.append(lines[i][1:-1].split('  '))
        os.chdir(self.basePath)
        added = self.formatter(added)
        deleted = self.formatter(deleted)

        # construct diff
        matchRes = match_ordered(deleted, added, self.comparer) if self.ordered else match_unordered(
            deleted, added, self.comparer)
        print(matchRes)
        diff: Dict[str, List] = {'former': matchRes['matched1'], 'latter': matchRes['matched2'], 'add': matchRes['unique2'],
                                 'delete': matchRes['unique1']}

        # item['addition'] means type when data type is 'add' for 'delete', otherwise means index in their own type sequence
        formerCount, latterCount = -1, 1
        res = []
        for key, values in diff.items():
            for value in values:
                if key == 'former':
                    value['addition'] = formerCount
                    formerCount -= 1
                elif key == 'latter':
                    value['addition'] = latterCount
                    latterCount += 1
                else:
                    value['addition'] = key
                res.append(value)
        return sorted(res, key=self.sortKey, reverse=True)


obs_observatory.py
from collections import OrderedDict
from flask import Flask, jsonify, request, Response
import os
import pandas as pd
from typing import List, Dict, Callable, Tuple
import json


app = Flask(__name__)
app.config['JSON_SORT_KEYS'] = False


BASEURL = os.getcwd() + '/oerv_obsdata'
PROPERTIES = ['Project Name', 'Repo Name', 'Success Rate',
              'Total', 'Succeeded', 'Failed', 'Unresolvable']

def formatData(rawData):
    res = []
    data = [[row[0], row[1], round(float(row[3]) / (int(row[2]) if row[2] != '0' else 1), 2), int(
        row[2]), int(row[3]), int(row[4]), int(row[5])] for row in rawData]
    data.sort(key=lambda row: row[3], reverse=True)
    for i in range(len(data)):
        res.append({PROPERTIES[j]: data[i][j] for j in range(len(PROPERTIES))})
    return res

def update(date_after: str = '2022-07-26') -> None:  # the short line is required
    '''
    get lists of commit dates and commit ids
    '''
    os.system('git pull origin master')
    os.system(
        'git log --pretty=format:"%cd,%H,%s" --date=format:"%Y-%m-%d" --after="{}" > ./commitlog.txt'.format(date_after))
    dates: List[str] = []
    commitID: List[str] = []
    with open('./commitlog.txt') as f:
        for line in f.readlines():
            temp = line[:-1].split(',')
            if temp[0] not in dates and temp[2][:2] == '20':
                dates.append(temp[0])
                commitID.append(temp[1])
    df = pd.DataFrame({'date': dates, 'id': commitID})
    df.to_csv('./date_id.csv', index=False)

def getSummary(id: str) -> object:
    os.system('git checkout {} ./obsData/projectStatistics.txt'.format(id))

    with open('./obsData/projectStatistics.txt') as f:
        rawData = list(map(lambda line: line[:-1].split('  '), f.readlines()))
    
    classData = {line[0] + '  ' + line[1]: {'total': int(line[2]), 'success': int(line[3])} for line in rawData}

    summary = []    
    classification = {'Core': ['openEuler:Mainline  standard'], 'Extend': ['openEuler:Epol  standard', 'openEuler:Factory  standard', 'Factory:RISC-V  22.09'], 'Third': []}
    
    for category in list(classification.keys()):
        temp = {'category': category, 'total': 0, 'success': 0, 'successRate': 0.0}
        for project in classification[category]:
            temp['total'] += classData[project]['total']
            temp['success'] += classData[project]['success']
        temp['successRate'] = round(temp['success'] / temp['total'], 2) if temp['total'] else 0.0
        summary.append(temp)
            
    return summary

@app.route('/getProjectStatistics')
def getProjectStatistics():    
    os.chdir(BASEURL)
    with open('./obsData/projectStatistics.txt') as f:
        rawData = list(map(lambda line: line[:-1].split('  '), f.readlines()))
    return jsonify({'data': formatData(rawData)})

@app.route('/getTrend')
def getTrend():
    os.chdir(BASEURL)
    df = pd.read_csv('./date_id.csv', index_col='date')
    dates = list(df.index)[::-1]
    res = []
    for date in dates:
        temp = {'date': date, 'data': [], 'total': 0}
        temp['data'] = getSummary(df.loc[date, 'id'])
        for summary in temp['data']:
            temp['total'] += summary['total']
        res.append(temp)
    return jsonify({'data': res})

@app.route('/getDiff', methods=['GET', 'POST'])
def getDiff():
    def compare(a: Dict, b: Dict):
        return a['Project Name'] == b['Project Name'] and a['Repo Name'] == b['Repo Name']

    def match(seq1: List, seq2: List, compare: Callable) -> Tuple[List, List, List, List]:
        '''
        given two sequence, find the matching items and remaining elements based on compare function
        this is a string match problem
        return (seq1, seq2, former, latter)
        Another traverse method: arr[:], see https://www.cnblogs.com/HZL2017/p/8894215.html
        '''
        matched1, matched2 = [], []
        mark = len(seq2) - 1
        for i in range(len(seq1) - 1, -1, -1):
            for j in range(mark, -1, -1):
                if compare(seq1[i], seq2[j]):
                    matched1.append(seq1.pop(i))
                    matched2.append(seq2.pop(j))
                    mark = j - 1
                    break
        return (seq1, seq2, matched1, matched2)

    def diff(id1: str, id2: str) -> Dict:
        os.system(
            'git diff {} {} ./obsData/projectStatistics.txt > {}'.format(id1, id2, './diff.txt'))
        # read data to variables deleted and added
        deleted: List = []
        added: List = []
        with open('./diff.txt') as f:
            lines = f.readlines()
            # start from 6th line
            for i in range(6, len(lines)):
                if lines[i][0] == '-':
                    # strip off + / - and new line character
                    deleted.append(lines[i][1:-1].split('  '))
                elif lines[i][0] == '+':
                    added.append(lines[i][1:-1].split('  '))
        added = formatData(added)
        deleted = formatData(deleted)

        # construct diff
        matched: Tuple[List, List, List, List] = match(deleted, added, compare)
        diff: Dict[str, List] = {'former': matched[2], 'latter': matched[3], 'add': matched[0],
                                 'delete': matched[1]}

        # item['addition'] means type when data type is 'add' for 'delete', otherwise means index in their own type sequence
        formerCount, latterCount = -1, 1
        res = []
        for key, values in diff.items():
            for value in values:
                if key == 'former':
                    value['addition'] = formerCount
                    formerCount -= 1
                elif key == 'latter':
                    value['addition'] = latterCount
                    latterCount += 1
                else:
                    value['addition'] = key
                res.append(value)
        return sorted(res, key= lambda line: line['Total'], reverse=True)

    os.chdir(BASEURL)
    df = pd.read_csv('./date_id.csv', index_col='date')
    data = request.values
    
    dates = list(df.index)
    date1 = dates[3]
    date2 = dates[0]
    id1, id2 = '', ''
    if len(data):
        date1 = data.get('date1')
        date2 = data.get('date2')
        id1, id2 = df.loc[date1, 'id'], df.loc[date2, 'id']
    else:
        id1, id2 = df.iloc[3, 0], df.iloc[0, 0]

    
    options = [{'label': date, 'value': date} for date in dates]

    return jsonify({'data': diff(id1, id2), 'date1': date1, 'date2': date2, 'options': options})


if __name__ == '__main__':
    os.chdir(BASEURL)
    update()
    os.chdir(BASEURL + '/..')

    app.debug = True
    app.run('0.0.0.0', '8080')


ObservatoryProject.py
from flask import Blueprint, jsonify, request
import os
import pandas as pd
from typing import List, Dict, Callable, Tuple

appObservatoryProject = Blueprint('appObservatoryProject', __name__)


BASEURL = os.getcwd() + '/oerv_obsdata'
PROPERTIES = ['Project Name', 'Repo Name', 'Success Rate',
              'Total', 'Succeeded', 'Failed', 'Unresolvable']

def formatData(rawData):
    res = []
    data = [[row[0], row[1], round(float(row[3]) / (int(row[2]) if row[2] != '0' else 1), 2), int(
        row[2]), int(row[3]), int(row[4]), int(row[5])] for row in rawData]
    data.sort(key=lambda row: row[3], reverse=True)
    for i in range(len(data)):
        res.append({PROPERTIES[j]: data[i][j] for j in range(len(PROPERTIES))})
    return res

def getSummary(id: str) -> object:
    os.system('git checkout {} ./obsData/projectStatistics.txt'.format(id))

    with open('./obsData/projectStatistics.txt') as f:
        rawData = list(map(lambda line: line[:-1].split('  '), f.readlines()))
    
    classData = {line[0] + '  ' + line[1]: {'total': int(line[2]), 'success': int(line[3])} for line in rawData}

    summary = []    
    classification = {'Core': ['openEuler:Mainline  standard'], 'Extend': ['openEuler:Epol  standard', 'openEuler:Factory  standard', 'Factory:RISC-V  22.09'], 'Third': []}
    
    for category in list(classification.keys()):
        temp = {'category': category, 'total': 0, 'success': 0, 'successRate': 0.0}
        for project in classification[category]:
            temp['total'] += classData[project]['total']
            temp['success'] += classData[project]['success']
        temp['successRate'] = round(temp['success'] / temp['total'], 2) if temp['total'] else 0.0
        summary.append(temp)
            
    return summary

@appObservatoryProject.route('/getProjectStatistics')
def getProjectStatistics():    
    with open('./oerv_obsdata/obsData/projectStatistics.txt') as f:
        rawData = list(map(lambda line: line[:-1].split('  '), f.readlines()))
    
    return jsonify({'data': formatData(rawData)})

@appObservatoryProject.route('/getTrend')
def getTrend():
    os.chdir(BASEURL)
    df = pd.read_csv('./date_id.csv', index_col='date')
    dates = list(df.index)[::-1]
    res = []
    for date in dates:
        temp = {'date': date, 'data': [], 'total': 0}
        temp['data'] = getSummary(df.loc[date, 'id'])
        for summary in temp['data']:
            temp['total'] += summary['total']
        res.append(temp)
    os.chdir(BASEURL + '/..')
    return jsonify({'data': res})

@appObservatoryProject.route('/getDiff', methods=['GET', 'POST'])
def getDiff():
    def compare(a: Dict, b: Dict):
        return a['Project Name'] == b['Project Name'] and a['Repo Name'] == b['Repo Name']

    def match(seq1: List, seq2: List, compare: Callable) -> Tuple[List, List, List, List]:
        '''
        given two sequence, find the matching items and remaining elements based on compare function
        this is a string match problem
        return (seq1, seq2, former, latter)
        Another traverse method: arr[:], see https://www.cnblogs.com/HZL2017/p/8894215.html
        '''
        matched1, matched2 = [], []
        mark = len(seq2) - 1
        for i in range(len(seq1) - 1, -1, -1):
            for j in range(mark, -1, -1):
                if compare(seq1[i], seq2[j]):
                    matched1.append(seq1.pop(i))
                    matched2.append(seq2.pop(j))
                    mark = j - 1
                    break
        return (seq1, seq2, matched1, matched2)

    def diff(id1: str, id2: str) -> Dict:
        os.system(
            'git diff {} {} ./obsData/projectStatistics.txt > {}'.format(id1, id2, './diff.txt'))
        # read data to variables deleted and added
        deleted: List = []
        added: List = []
        with open('./diff.txt') as f:
            lines = f.readlines()
            # start from 6th line
            for i in range(6, len(lines)):
                if lines[i][0] == '-':
                    # strip off + / - and new line character
                    deleted.append(lines[i][1:-1].split('  '))
                elif lines[i][0] == '+':
                    added.append(lines[i][1:-1].split('  '))
        added = formatData(added)
        deleted = formatData(deleted)

        # construct diff
        matched: Tuple[List, List, List, List] = match(deleted, added, compare)
        diff: Dict[str, List] = {'former': matched[2], 'latter': matched[3], 'add': matched[0],
                                 'delete': matched[1]}

        # item['addition'] means type when data type is 'add' for 'delete', otherwise means index in their own type sequence
        formerCount, latterCount = -1, 1
        res = []
        for key, values in diff.items():
            for value in values:
                if key == 'former':
                    value['addition'] = formerCount
                    formerCount -= 1
                elif key == 'latter':
                    value['addition'] = latterCount
                    latterCount += 1
                else:
                    value['addition'] = key
                res.append(value)
        return sorted(res, key= lambda line: line['Total'], reverse=True)

    os.chdir(BASEURL)
    df = pd.read_csv('./date_id.csv', index_col='date')
    
    data = request.values
    
    dates = list(df.index)
    date1 = dates[3]
    date2 = dates[0]
    id1, id2 = '', ''
    if len(data):
        date1 = data.get('date1')
        date2 = data.get('date2')
        id1, id2 = df.loc[date1, 'id'], df.loc[date2, 'id']
    else:
        id1, id2 = df.iloc[3, 0], df.iloc[0, 0]

    options = [{'label': date, 'value': date} for date in dates]
    diffRes = diff(id1, id2)
    os.chdir(BASEURL + '/..')
    return jsonify({'data': diffRes, 'date1': date1, 'date2': date2, 'options': options})


ObservatoryPackages.py
from struct import pack
from flask import Blueprint, request, jsonify
from typing import Dict, List
import os
import re
import pandas as pd
import requests

from Utils import Differ

appObservatoryPackages = Blueprint('appObservatoryPackages', __name__)


@appObservatoryPackages.route('/search-package', methods=['POST'])
def searchPackage():
    '''
    :return [(projectName, packageName), ...]
    '''
    def constructURL(target):
        base_url = 'https://build.tarsier-infra.com/search?'
        args = 'utf8=✓&search_text={}&search_for=0&name=1&attrib_type_id='.format(
            target)
        return base_url + args

    # get html text
    target = request.values.get('packageName', 'gcc')
    account = {'username': 'cyanic', 'password': '123456'}
    response = requests.get(constructURL(
        target), auth=requests.auth.HTTPBasicAuth(**account))
    text = response.text

    # filter out all related path in form of [(projectName, packageName), ...]
    raw = re.findall(r'package .*?title="(.*?)".*?title="(.*?)"', text)

    with open('./oerv_obsdata/obsData/projectStatistics.txt') as f1:
        projectsNames = set(re.findall(r'^|\n(.*?)  ', f1.read())[1:])
        print(projectsNames)

    res = list(filter(lambda name: name[0] in projectsNames, raw))
    res = [{'projectName': item[0], 'packageName': item[1]} for item in res]
    print('res', res)
    return jsonify({'data': res})


@appObservatoryPackages.route('/getRepositories')
def getRepositories():
    with open('./oerv_obsdata/obsData/projectStatistics.txt') as f:
        namePairs = re.findall(r'^|\n(.*?)  (.*?)  ', f.read())[1:]
    res = {}
    for pair in namePairs:
        if pair[0] not in res:
            res[pair[0]] = [pair[1]]
        else:
            res[pair[0]].append(pair[1])
    res = [{'value': key, 'label': key, 'children': [{'value': v, 'label': v}
                                                     for v in val]} for key, val in res.items()]
    return jsonify({'data': res})


@appObservatoryPackages.route('/getPackagesDiffSpatial', methods=['POST'])
def getPackagesDiffSpatial():

    def diff(path1, path2) -> Dict[str, List[Dict[str, str]]]:
        def extract(path) -> Dict[str, str]:
            f = open(path)
            data = {pair[0]: pair[1] for pair in list(
                map(lambda s: s.split('  '), f.read().split('\n')))[:-1]}
            f.close()
            return data

        data1 = extract(path1)
        data2 = extract(path2)

        packageNames1 = set(data1.keys())
        packageNames2 = set(data2.keys())
        intersection = packageNames1.intersection(packageNames2)
        unique1 = [{'packageName': packageName, 'status': data1[packageName]}
                   for packageName in packageNames1.difference(packageNames2)]
        unique2 = [{'packageName': packageName, 'status': data2[packageName]}
                   for packageName in packageNames2.difference(packageNames1)]

        same = []
        diff = []
        for packageName in intersection:
            status1, status2 = data1[packageName], data2[packageName]
            if status1 == status2:
                same.append({'packageName': packageName, 'status': status1})
            else:
                diff.append({'packageName': packageName,
                            'status1': status1, 'status2': status2})
        return {'unique1': unique1, 'unique2': unique2, 'same': same, 'diff': diff}

    # repositoryName1, repositoryName2 = request.values.values
    repositoryName1, repositoryName2 = request.values.get(
        'repositoryName1'), request.values.get('repositoryName2')
    path1 = './oerv_obsdata/obsData/' + \
        repositoryName1.replace(':', '+') + '.list'
    path2 = './oerv_obsdata/obsData/' + \
        repositoryName2.replace(':', '+') + '.list'
    res = diff(path1, path2)
    return jsonify(res)


@appObservatoryPackages.route('/getDates', methods=['POST'])
def getDates():
    path = './oerv_obsdata/timeline/' + \
        request.values.get('repositoryName').replace(':', '+') + '.txt'
    dates = []
    with open(path) as f:
        for line in f.readlines():
            date = line[:10]
            if date not in dates:
                dates.append(date)
    dates.reverse()
    return jsonify({'dates': dates})


@appObservatoryPackages.route('/getPackagesDiffTemporal', methods=['POST'])
def getPackagesDiffTemporal():
    basePath = os.getcwd()
    repositoryPath = basePath + '/oerv_obsdata'
    filePath = './obsData/' + \
        request.values.get('repositoryName').replace(':', '+') + '.list'

    # get ids
    df = pd.read_csv('./oerv_obsdata/date_id.csv', index_col='date')
    data = request.values
    date1 = data.get('date1')
    date2 = data.get('date2')
    # print(date1 in list(df.index))
    id1, id2 = df.loc[date1, 'id'], df.loc[date2, 'id']

    def sortKey(item): return item['Package Name']

    def formatter(data): return [
        {'Package Name': item[0], 'Status': item[1]} for item in data]
    compare = ['Package Name']
    differ = Differ(id1, id2, sortKey, formatter, False,
                    compare, filePath, repositoryPath, basePath)
    diffRes = differ.gitDiff()

    return jsonify({'data': diffRes})


ObservatoryProjects.py
from flask import Blueprint, jsonify, request
import os
import pandas as pd
from typing import List, Dict, Tuple, Callable
from Utils import Differ, match_ordered

appObservatoryProject = Blueprint('appObservatoryProject', __name__)


basePath = os.getcwd()
PROPERTIES = ['Project Name', 'Repo Name', 'Success Rate',
              'Total', 'Succeeded', 'Failed', 'Unresolvable']


def formatData(rawData):
    res = []
    data = [[row[0], row[1], round(float(row[3]) / (int(row[2]) if row[2] != '0' else 1), 2), int(
        row[2]), int(row[3]), int(row[4]), int(row[5])] for row in rawData]
    data.sort(key=lambda row: row[3], reverse=True)
    for i in range(len(data)):
        res.append({PROPERTIES[j]: data[i][j] for j in range(len(PROPERTIES))})
    return res


def getSummary(id: str) -> object:
    os.system('git checkout {} ./obsData/projectStatistics.txt'.format(id))

    with open('./obsData/projectStatistics.txt') as f:
        rawData = list(map(lambda line: line[:-1].split('  '), f.readlines()))

    classData = {line[0] + '  ' + line[1]: {'total': int(line[2]), 'success': int(line[3])} for line in rawData}

    summary = []
    classification = {'Core': ['openEuler:Mainline  standard'], 'Extend': [
        'openEuler:Epol  standard', 'openEuler:Factory  standard', 'Factory:RISC-V  22.09'], 'Third': []}

    for category in list(classification.keys()):
        temp = {'category': category, 'total': 0,
                'success': 0, 'successRate': 0.0}
        for project in classification[category]:
            temp['total'] += classData[project]['total']
            temp['success'] += classData[project]['success']
        temp['successRate'] = round(
            temp['success'] / temp['total'], 2) if temp['total'] else 0.0
        summary.append(temp)

    return summary


@appObservatoryProject.route('/getProjectStatistics')
def getProjectStatistics():
    with open('./oerv_obsdata/obsData/projectStatistics.txt') as f:
        rawData = list(map(lambda line: line[:-1].split('  '), f.readlines()))

    return jsonify({'data': formatData(rawData)})


@appObservatoryProject.route('/getTrend')
def getTrend():
    df = pd.read_csv('./oerv_obsdata/date_id.csv', index_col='date')
    dates = list(df.index)[::-1]
    res = []
    os.chdir(basePath + '/oerv_obsdata')
    for date in dates:
        temp = {'date': date, 'data': [], 'total': 0}
        temp['data'] = getSummary(df.loc[date, 'id'])
        for summary in temp['data']:
            temp['total'] += summary['total']
        res.append(temp)
    os.chdir(basePath)
    return jsonify({'data': res})


@appObservatoryProject.route('/getDiff', methods=['GET', 'POST'])
def getDiff():
    def compare(a: Dict, b: Dict):
        return a['Project Name'] == b['Project Name'] and a['Repo Name'] == b['Repo Name']

    # get commit id based on dates
    df = pd.read_csv('./oerv_obsdata/date_id.csv', index_col='date')
    data = request.values
    dates = list(df.index)
    date1 = dates[3]
    date2 = dates[0]
    id1, id2 = '', ''
    if len(data):
        date1 = data.get('date1')
        date2 = data.get('date2')
        id1, id2 = df.loc[date1, 'id'], df.loc[date2, 'id']
    else:
        id1, id2 = df.iloc[3, 0], df.iloc[0, 0]

    filePath = './obsData/projectStatistics.txt'
    repositoryPath = basePath + '/oerv_obsdata'
    def sortKey(line): return line['Total']
    differ = Differ(id1, id2, sortKey, formatData, True,
                    compare, filePath, repositoryPath, basePath)
    diffRes = differ.gitDiff()

    options = [{'label': date, 'value': date} for date in dates]
    return jsonify({'data': diffRes, 'date1': date1, 'date2': date2, 'options': options})


Update.py
from flask import Blueprint, Response
import os
import pandas as pd
from typing import List


# import requests
# from requests.auth import HTTPBasicAuth
import re 

appUpdate = Blueprint('appUpdate', __name__)

BASEURL = os.getcwd()

@appUpdate.route('/updateLocal')
def updateLocal(date_after: str = '2022-07-26') -> None:  # the short line is required
    '''
    - Get lists of commit dates and commit ids
    - Generate a file that contains the names of all related projects
    '''
    os.chdir(BASEURL + '/oerv_obsdata')

    # Part I. Generate a csv file that contains information about dates and commit ids
    os.system('git pull origin master')
    os.system(
        'git log --pretty=format:"%cd,%H,%s" --date=format:"%Y-%m-%d" --after="{}" > ./commitlog.txt'.format(date_after))
    dates: List[str] = []
    commitID: List[str] = []
    with open('./commitlog.txt') as f:
        for line in f.readlines():
            temp = line[:-1].split(',')
            if temp[0] not in dates and temp[2][:2] == '20':
                dates.append(temp[0])
                commitID.append(temp[1])
    df = pd.DataFrame({'date': dates, 'id': commitID})
    df.to_csv('./date_id.csv', index=False)

    # Part II. Generate a txt file that contains information about the names of all related projects
    # with open('./obsData/projectStatistics.txt') as f1:
    #     with open('./projectsNames.txt', 'w') as f2:
    #         raw = f1.read()
    #         names = re.findall(r'^|\n(.*?)  (.*?)  ', raw)[1:]
    #         f2.write('\n'.join(['{},{}'.format(name[0], name[1]) for name in names]))

    os.chdir(BASEURL)
    return Response('Update Local Success')

@appUpdate.route('/updateRepository')
def updateRepository():
    # pull from remote
    os.chdir(BASEURL + '/oerv_script/oerv_obsdata')
    os.system('git pull --no-ff')

    os.chdir(BASEURL + '/oerv_script/script/')
    os.system('python main.py')
    os.chdir(BASEURL)   
    return Response('Update repository success')